File: 992. Subarrays with K Different Integers.java
class Solution {
    public int subarraysWithKDistinct(int[] nums, int k) {
        return helper(nums, k) - helper(nums, k-1);
    }
    private int helper(int[] nums, int k) {
        if (k > nums.length || k == 0) return 0;
        Map<Integer, Integer> counts = new HashMap<>();
        int left = 0;
        int count = 0;
        for (int right = 0; right < nums.length; right++) {
            counts.put(nums[right], counts.getOrDefault(nums[right], 0) + 1);
            while (left < right && counts.size() > k) {
                counts.put(nums[left], counts.get(nums[left]) - 1);
                if (counts.get(nums[left]) == 0) counts.remove(nums[left]);
                left++;
            }
            count += right - left + 1;
        }
        System.out.println(count);
        return count;
    }
}

File: python/992. Subarrays with K Different Integers.py
class Solution:
    def subarraysWithKDistinct(self, nums: List[int], k: int) -> int:
        def find(nums, k):
            freq = {}
            unique, answer, left = 0, 0, 0
            for right in range(len(nums)):
                if nums[right] not in freq:
                    freq[nums[right]] = 0
                freq[nums[right]] += 1
                if freq[nums[right]] == 1:
                    unique += 1
                while left <= right and unique > k:
                    freq[nums[left]] -= 1
                    if freq[nums[left]] == 0:
                        unique -= 1
                    left += 1
                answer += right - left + 1
            return answer
        return find(nums, k) - find(nums, k-1)

